File: backend/models/project.py
"""
Python mirror of shared/models/project.ts
Used for backend validation and engine consumption.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional, Literal
from datetime import datetime


GameMode = Literal['rpg', 'galgame']


@dataclass
class ProjectMeta:
    name: str
    mode: GameMode
    version: str = '0.1.0'
    created_at: str = ''
    updated_at: str = ''
    description: str = ''

    def __post_init__(self):
        now = datetime.now().isoformat()
        if not self.created_at:
            self.created_at = now
        if not self.updated_at:
            self.updated_at = now


AssetType = Literal['sprite', 'tileset', 'background', 'portrait', 'effect', 'audio', 'ui', 'other']


@dataclass
class AssetRef:
    id: str
    name: str
    type: AssetType
    relative_path: str
    width: int = 0
    height: int = 0
    imported_at: str = ''


@dataclass
class UIColorTheme:
    primary: str = '#89b4fa'
    secondary: str = '#a6e3a1'
    background: str = '#1e1e2e'
    surface: str = '#313244'
    text: str = '#cdd6f4'
    text_muted: str = '#6c7086'
    accent: str = '#f5c2e7'
    danger: str = '#f38ba8'


@dataclass
class GameProject:
    meta: ProjectMeta
    assets: list[AssetRef] = field(default_factory=list)
    ui: dict = field(default_factory=dict)
    rpg: Optional[dict] = None
    galgame: Optional[dict] = None

    @classmethod
    def from_dict(cls, data: dict) -> GameProject:
        """Deserialize from JSON dict (matches TS GameProject)."""
        meta = ProjectMeta(
            name=data['meta']['name'],
            mode=data['meta']['mode'],
            version=data['meta'].get('version', '0.1.0'),
            created_at=data['meta'].get('createdAt', ''),
            updated_at=data['meta'].get('updatedAt', ''),
            description=data['meta'].get('description', '')
        )
        assets = [
            AssetRef(
                id=a['id'],
                name=a['name'],
                type=a['type'],
                relative_path=a['relativePath'],
                width=a.get('width', 0),
                height=a.get('height', 0),
                imported_at=a.get('importedAt', '')
            )
            for a in data.get('assets', [])
        ]
        return cls(
            meta=meta,
            assets=assets,
            ui=data.get('ui', {}),
            rpg=data.get('rpg'),
            galgame=data.get('galgame')
        )

    def to_dict(self) -> dict:
        """Serialize to JSON-compatible dict."""
        return {
            'meta': {
                'name': self.meta.name,
                'mode': self.meta.mode,
                'version': self.meta.version,
                'createdAt': self.meta.created_at,
                'updatedAt': self.meta.updated_at,
                'description': self.meta.description
            },
            'assets': [
                {
                    'id': a.id,
                    'name': a.name,
                    'type': a.type,
                    'relativePath': a.relative_path,
                    'width': a.width,
                    'height': a.height,
                    'importedAt': a.imported_at
                }
                for a in self.assets
            ],
            'ui': self.ui,
            'rpg': self.rpg,
            'galgame': self.galgame
        }

File: shared/models/project.ts
// ── Shared Project Data Models ────────────────────────────
// Used by both frontend (TypeScript) and backend (Python)
// Python side mirrors this in backend/models/project.py

export type GameMode = 'rpg' | 'galgame'

export interface ProjectMeta {
  name: string
  mode: GameMode
  version: string
  createdAt: string
  updatedAt: string
  description: string
}

// ── Asset ─────────────────────────────────────────────────

export type AssetType = 'sprite' | 'tileset' | 'background' | 'portrait' | 'effect' | 'audio' | 'ui' | 'other'

export interface AssetRef {
  id: string
  name: string
  type: AssetType
  relativePath: string
  dataUrl?: string      // base64 data URL for embedded assets
  width: number
  height: number
  importedAt: string
}

// ── RPG Mode ──────────────────────────────────────────────

export interface TileLayer {
  id: string
  name: string
  data: number[][]     // [row][col] tile indices
  visible: boolean
  opacity: number
  isCollision: boolean
}

export interface RPGMap {
  id: string
  name: string
  width: number        // in tiles
  height: number
  tileWidth: number    // pixel size per tile (default 32)
  tileHeight: number
  tilesetId: string    // reference to AssetRef.id
  layers: TileLayer[]
}

export interface RPGEntity {
  id: string
  type: 'player_spawn' | 'npc' | 'portal' | 'trigger' | 'item'
  name: string
  spriteId: string
  x: number
  y: number
  properties: Record<string, unknown>
}

export interface RPGData {
  maps: RPGMap[]
  entities: RPGEntity[]
  currentMapId: string
}

// ── Galgame Mode ──────────────────────────────────────────

export interface DialogueNode {
  id: string
  speakerName: string
  speakerPortraitId: string
  portraitExpression: string   // e.g. 'neutral', 'happy', 'sad'
  text: string
  nextNodeId: string | null
  effects: string[]             // e.g. ['fadeIn', 'shake']
  sceneId?: string              // bound scene
  changeSceneId?: string        // switch to this scene after this dialogue
  transitionEffect?: string     // transition when changing scene (overrides scene default)
}

export interface BranchNode {
  id: string
  prompt: string
  choices: ChoiceOption[]
  sceneId?: string              // bound scene
}

export interface ChoiceOption {
  text: string
  nextNodeId: string
  condition?: string            // e.g. 'affection >= 10'
}

export interface SceneDef {
  id: string
  name: string
  backgroundType?: 'image' | 'video'  // background asset type (default: 'image')
  backgroundPath?: string             // file path to background image/video
  bgmPath?: string | null             // file path to BGM audio
  transition?: string                 // preset: 'none','fade','slideLeft','slideRight','zoomIn','zoomOut','dissolve'
  customTransition?: string           // custom transition name (takes priority over preset if set)
  dialogueNodeIds?: string[]          // dialogue/branch nodes bound to this scene
  characterIds?: string[]             // characters appearing in this scene
}

export interface CharacterDef {
  id: string
  name: string
  portraitPath: string               // default portrait (data URL or path)
  expressions: Record<string, string> // expression name → portrait path
  position?: 'left' | 'center' | 'right'  // screen position
}

export interface GalgameData {
  scenes: SceneDef[]
  dialogueNodes: DialogueNode[]
  branchNodes: BranchNode[]
  characters: CharacterDef[]
  variables: Record<string, number>  // e.g. { affection: 0, flags: {...} }
  startSceneId: string
  startNodeId: string
}

// ── UI Customization ──────────────────────────────────────

export interface UIColorTheme {
  primary: string
  secondary: string
  background: string
  surface: string
  text: string
  textMuted: string
  accent: string
  danger: string
}

export interface RPGHUDConfig {
  showHealth: boolean
  showMana: boolean
  showMinimap: boolean
  showInventory: boolean
  components: HUDComponent[]
}

export interface HUDComponent {
  id: string
  type: 'health' | 'mana' | 'minimap' | 'inventory' | 'quest'
  x: number
  y: number
  width: number
  height: number
  visible: boolean
}

export interface DialogueBoxConfig {
  position: 'bottom' | 'top' | 'fullscreen'
  backgroundStyle: 'solid' | 'gradient' | 'image'
  backgroundColor: string
  borderRadius: number
  fontSize: number
  textColor: string
  nameBoxColor: string
  typewriterSpeed: number       // ms per character
}

export interface MenuUIConfig {
  titleScreen: MenuScreenDef
  saveLoadScreen: MenuScreenDef
  settingsScreen: MenuScreenDef
}

export interface MenuScreenDef {
  backgroundId: string
  buttons: MenuButtonDef[]
}

export interface MenuButtonDef {
  id: string
  text: string
  x: number
  y: number
  width: number
  height: number
  style: 'primary' | 'secondary' | 'text'
}

export interface UICustomization {
  theme: UIColorTheme
  rpgHUD: RPGHUDConfig
  dialogueBox: DialogueBoxConfig
  menu: MenuUIConfig
  fontFamily: string
}

// ── Root Project ──────────────────────────────────────────

export interface GameProject {
  meta: ProjectMeta
  assets: AssetRef[]
  ui: UICustomization
  rpg: RPGData | null
  galgame: GalgameData | null
}
